ArduCopter/events.cpp

```diff
@@ -306,6 +306,77 @@ void Copter::gpsglitch_check()
     }
 }
 
+// executes lean-angle failsafe if lean angle is too high for longer than a set threshold
+// indicates the wind is too high to fly safely in planck mode
+void Copter::failsafe_lean_check(uint16_t duration_ms)
+{
+    // trigger with X seconds of failures while in a PLANCK mode
+    bool valid_mode = (control_mode == Mode::Number::PLANCKTRACK ||
+                       control_mode == Mode::Number::PLANCKRTB ||
+                       control_mode == Mode::Number::PLANCKLAND);
+
+    if (attitude_control->lean_angle()*100 > attitude_control->planck_lean_angle_max()){
+        if(failsafe.lean_high_time_ms<=0)
+        {
+            failsafe.lean_high_time_ms = AP_HAL::millis();
+        }
+    }
+    else{
+        failsafe.lean_high_time_ms = 0;
+    }
+
+    bool timeout = uint16_t(AP_HAL::millis()-failsafe.lean_high_time_ms) > g.planck_angle_fs_to;
+    bool trigger_event = valid_mode && timeout;
+
+    // check for clearing of event
+    if (trigger_event != failsafe.lean) {
+        if (trigger_event) {
+            failsafe_lean_on_event();
+        } else {
+            AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_LEAN, LogErrorCode::ERROR_RESOLVED);
+            failsafe.lean = false;
+        }
+    }
+}
+
+// terrain failsafe action
+void Copter::failsafe_lean_on_event()
+{
+    failsafe.lean = true;
+    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_LEAN, LogErrorCode::FAILSAFE_OCCURRED);
+
+    Failsafe_Action desired_action;
+    switch (g.failsafe_planck_angle) {
+            case FS_PLANCK_ANGLE_DISABLED:
+                desired_action = Failsafe_Action_None;
+                break;
+
+            case FS_PLANCK_ANGLE_ENABLED_PLANCK_TRACK_PLANCK_LAND:
+                gcs().send_text(MAV_SEVERITY_CRITICAL,"Failsafe: Lean angle high - Landing");
+                desired_action = Failsafe_Action_Planck_Track_Land;
+                break;
+
+            case FS_PLANCK_ANGLE_WARN:
+            default: // if an invalid parameter value is set, the fallback is RTL
+                gcs().send_text(MAV_SEVERITY_WARNING,"Failsafe: Lean angle high");
+                desired_action = Failsafe_Action_None;
```

Comment:
missing a break, nbd.

Reply:
where should the break be?  The FS_PLANCK_ANGLE_WARN case is meant to fall through.

Reply:
Yeah, generally its good practice to still put a `break` after the last/default case in a switch statement.  This is fine though.